core/src/test/scala/org/apache/spark/sql/delta/DeltaSourceSchemaEvolutionSuite.scala

```diff
@@ -0,0 +1,342 @@
+      StartStream(checkpointLocation = getDefaultCheckpoint.toString),
+      ProcessAllAvailable(),
+      // See how the schema returns one more dimension for `c`
+      CheckAnswer((-1 until 5).map(_.toString).map(i => (i, i, null)): _*)
+    )
+
+    // Cannot use schema with different partition schema than the table
+    val newSchemaWithPartition = PersistedSchema(log.tableId, 100,
+      new StructType().add("a", StringType, true)
+        .add("b", StringType, true),
+      new StructType().add("b", StringType, true))
+    schemaLog.evolveSchema(newSchemaWithPartition)
+
+    assert {
+      val e = intercept[DeltaAnalysisException] {
+        val q = readStream(schemaLocation = Some(getDefaultSchemaLocation.toString))
+          .writeStream
+          .option("checkpointLocation", getDefaultCheckpoint.toString)
+          .outputMode("append")
+          .format("console")
+          .start()
+        q.processAllAvailable()
+        q.stop()
+      }
+      ExceptionUtils.getRootCause(e).asInstanceOf[DeltaAnalysisException]
+        .getErrorClass == "DELTA_STREAMING_SCHEMA_LOG_INCOMPATIBLE_PARTITION_SCHEMA"
+    }
+
+    // Cannot use schema from another table
+    val newSchemaWithTableId = PersistedSchema(
+      "some_random_id", 100,
+      new StructType().add("a", StringType, true)
+        .add("b", StringType, true),
+      new StructType())
+    schemaLog.evolveSchema(newSchemaWithTableId)
+    assert {
+      val e = intercept[DeltaAnalysisException] {
+        val q = readStream(schemaLocation = Some(getDefaultSchemaLocation.toString))
+          .writeStream
+          .option("checkpointLocation", getDefaultCheckpoint.toString)
+          .outputMode("append")
+          .format("console")
+          .start()
+        q.processAllAvailable()
+        q.stop()
+      }
+      ExceptionUtils.getRootCause(e).asInstanceOf[DeltaAnalysisException]
+        .getErrorClass == "DELTA_STREAMING_SCHEMA_LOG_INCOMPATIBLE_DELTA_TABLE_ID"
+    }
+  }
+
+  test("concurrent schema log modification should be detected") {
+    withStarterTable { implicit log =>
+      // Note: this test assumes schema log files are written one after another, which is majority
+      // of the case; True concurrent execution would require commit service to protected against.
+      val schemaLocation = getDefaultSchemaLocation.toString
+      val snapshot = log.update()
+      val schemaLog1 = DeltaSourceSchemaLog.create(spark, schemaLocation, snapshot)
+      val schemaLog2 = DeltaSourceSchemaLog.create(spark, schemaLocation, snapshot)
+      val newSchema = PersistedSchema("1", 1, new StructType(), new StructType())
+
+      schemaLog1.evolveSchema(newSchema)
+      val e = intercept[DeltaAnalysisException] {
+        schemaLog2.evolveSchema(newSchema)
+      }
+      assert(e.getErrorClass == "DELTA_STREAMING_SCHEMA_LOCATION_CONFLICT")
+    }
+  }
+
+  // TODO: more tests
```

Comment:
what other tests do you plan to add? at a high-level

Reply:
Yep, so this is just part 1 to reduce the review overhead, in the next part, I will be adding the main logic for non-additive schema evolution and there would be a lot more tests!